### PR TITLE
Development

### PR DESCRIPTION
**Pull request by Leidse**

### CODE REVIEW

File: src/robot_arm.hpp [hunk 1]
```diff
@@ -1,19 +1,45 @@
...
+     * @param[int[3]] coordinates
+     * @param[int] speed
+     */
+    void move(int coordinates[3], int _speed);
+    /**
+     * @brief Return the current coordinates
+     *
+     * @return current_coordinates
+     */
+    int getCoordinates(char dimension);
```
Comment (vasilishomegas) [inline src/robot_arm.hpp]: Please document parameter

File: src/robot_arm.hpp [hunk 2]
```diff
@@ -1,19 +1,45 @@
...
-    void copy_data(int *array1, int *array2);
+    /**
+     * @brief Construct a new Robot Arm object
+     *
+     */
+    RobotArm();
+    /**
+     * @brief Move the arm to the set coordinates
+     *
+     *
```
Comment (vasilishomegas) [inline src/robot_arm.hpp]: Remove excess emptly lines.

File: documents/Robot Arm.simp.bak [hunk 3]
```diff
@@ -0,0 +1,271 @@
+﻿<?xml version="1.0" encoding="utf-8" standalone="yes"?>
```
Comment (vasilishomegas) [inline documents/Robot Arm.simp.bak]: This file shouldn't be here.

File: src/settings.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#include "settings.hpp"
+
+Settings::Settings(int motorSpeed) : motorSpeed(motorSpeed) {
+}
+
+void Settings::setMotorSpeed(int speed) {
+    motorSpeed = speed;
+    // hwlib::cout << "Set motorspeed to: " << motorSpeed << hwlib::endl;
```
Comment (vasilishomegas) [inline src/settings.cpp]: If this line is no longer required, delete it.

File: src/main.cpp [hunk 5]
```diff
@@ -1,15 +1,25 @@
...
 #include "wrap-hwlib.hpp"
 
 int main() {
     WDT->WDT_MR = WDT_MR_WDDIS;
-    int coordinates[3] = {4, 12, 5};
-
     hwlib::wait_ms(1000);
 
-    robot_arm uarm_swift_pro(hwlib::target::pin_out(hwlib::target::pins::d7));
+    int coordinates[3] = {150, 150, 150};
```
Comment (vasilishomegas) [inline src/main.cpp]: Don't use C-style arrays

File: src/robot_arm.cpp [hunk 6]
```diff
@@ -1,19 +1,69 @@
...
+ * @license   MIT
+ */
 #include "robot_arm.hpp"
 
-void robot_arm::move(int coordinates[3]) {
-    copy_data(coordinates, goto_coordinates);
+RobotArm::RobotArm() : conn(115200, UARTController::ONE) {
+}
+
+void RobotArm::move(int coordinates[3], int _speed) {
```
Comment (vasilishomegas) [inline src/robot_arm.cpp]: C-style array

File: src/robot_arm.cpp [hunk 7]
```diff
@@ -1,19 +1,69 @@
...
 
-void robot_arm::copy_data(int *array1, int *array2) {
-    for (int i = 0; i < (sizeof(array1) / 4); i++) {
-        array2[i] = array1[i];
+void RobotArm::executeAction(Actions action) {
+    action = action;
+    hwlib::cout << "Executing Action" << hwlib::endl;
+}
+
+void RobotArm::determineGCode(int coordinates[3]) {
```
Comment (vasilishomegas) [inline src/robot_arm.cpp]: C-style array

File: src/robot_arm.hpp [hunk 8]
```diff
@@ -1,19 +1,89 @@
...
-class robot_arm {
+class RobotArm {
   private:
+    enum class Actions { reset, calibrate };
+
+    Actions action;
+
+    int speed;
     int current_coordinates[3];
     int goto_coordinates[3];
```
Comment (vasilishomegas) [inline src/robot_arm.hpp]: C-style arrays

File: src/robot_arm.hpp [hunk 9]
```diff
@@ -1,19 +1,89 @@
...
+     */
+    RobotArm();
+
+    /**
+     * @brief Move the arm to the set coordinates
+     *
+     * This function will take an array existing of 3 integer values and an integer that resembles the speed. These parameters will
+     * me transformed into a G-Code. This will be send to the uArm using the uart_connection class.
+     *
+     * @param[int[3]] coordinates
```
Comment (vasilishomegas) [inline src/robot_arm.hpp]: Check wiki for proper doxygen format. This is not how to document parameters.

File: src/robot_arm.hpp [hunk 10]
```diff
@@ -1,19 +1,89 @@
...
+     *
+     * @return current_coordinates
+     */
+    int getCoordinates(char dimension);
+    /**
+     * @brief Determine G-Code for a desired location.
+     *
+     *
+     *
+     * @param[int[3]] coordinates
```
Comment (vasilishomegas) [inline src/robot_arm.hpp]: Check wiki for proper doxygen documentation

File: src/robot_arm.hpp [hunk 11]
```diff
@@ -1,19 +1,89 @@
...
+     */
+    void determineGCode(int coordinates[3]);
+    /**
+     * @brief Determine G-Code for a desired action.
+     *
+     *
+     *
+     * @param[Actions]] action
```
Comment (vasilishomegas) [inline src/robot_arm.hpp]: Same here, check the doxygen of all functions (I won't add a comment for each).

File: src/uart_connection.cpp [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    USARTControllerInitialized = true;
+}
+
+unsigned int UARTConnection::available() {
+    if (!USARTControllerInitialized) {
+        return false;
+    }
+
+    /// We use the USART Channel status register to check if there is data available in the receiver buffer.
+    // return (hardwareUSART->US_CSR & 1) != 0;
```
Comment (vasilishomegas) [inline src/uart_connection.cpp:87]: Remove unused code

File: src/uart_connection.cpp [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+
+    if ((hardwareUSART->US_CSR & 1) != 0) {
+        // rxBuffer[rxBufferIndex++] = receiveByte();
```
Comment (vasilishomegas) [inline src/uart_connection.cpp:90]: Remove unused code

File: src/uart_connection.cpp [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+    return true;
+}
+
+bool UARTConnection::send(const char *str) {
+    if (!USARTControllerInitialized) {
+        return false;
+    }
+
+    for (const char *p = str; *p != '\0'; p++) {
```
Comment (vasilishomegas) [inline src/uart_connection.cpp:112]: C-style

File: src/uart_connection.cpp [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        sendByte(*p);
+    }
+
+    return true;
+}
+
+bool UARTConnection::send(const char *data, size_t length) {
```
Comment (vasilishomegas) [inline src/uart_connection.cpp:119]: C-style

File: src/robot_arm.cpp [hunk 16]
```diff
@@ -1,19 +1,69 @@
...
+    hwlib::cout << "Moving to " << coordinates[0] << ", " << coordinates[1] << ", " << coordinates[2] << "." << hwlib::endl;
+}
+
+void RobotArm::determineGCode(Actions action) {
+    switch (action) {
```
Comment (vasilishomegas) [inline src/robot_arm.cpp]: Why do you use a switch? Also, why do you not break the cases? If this is on purpose, document it.

File: src/robot_arm.cpp [hunk 17]
```diff
@@ -1,19 +1,69 @@
...
+    case Actions::reset:
+        hwlib::cout << "Resetting" << hwlib::endl;
+    case Actions::calibrate:
+        hwlib::cout << "Calibrating" << hwlib::endl;
+    default:
+        hwlib::cout << "This isn't a legit move" << hwlib::endl;
+    }
+}
+
+int RobotArm::getCoordinates(char dimension) {
```
Comment (vasilishomegas) [inline src/robot_arm.cpp]: Why a switch before and now if's? On another note, you return but still use else, although you'll never get there anyway.

**Review by vasilishomegas — CHANGES_REQUESTED** [review]

**Comment by vasilishomegas** [comment]
There are several merge conflicts that need fixing.

File: BuildModule.cmake [hunk 18]
```diff
@@ -0,0 +1,19 @@
+include (../../flags.cmake)
```
Comment (JulianvDoorn) [inline BuildModule.cmake:1]: Undocumented feature but it is preferred to replace this with `include (${build_environment}/flags.cmake)`

File: CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+include (Sources.cmake)
+
+if (NOT ${test_build})
+set (CMAKE_TOOLCHAIN_FILE ${toolchain}/compilers/arm-none-eabi.cmake)
+else (NOT ${test_build})
+set (CMAKE_TOOLCHAIN_FILE ${toolchain}/compilers/host-gcc.cmake)
+endif (NOT ${test_build})
+
+project (arduino-template CXX)
```
Comment (JulianvDoorn) [inline CMakeLists.txt:13]: Rename arduino-template to robot-arm

File: Sources.cmake [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+include_directories (${catch}/single_include)
+
+# Source Files:
+
+set (sources
+    src/coordinate3d.cpp
+    src/libc-stub.cpp
+    src/main.cpp
+    src/robot_arm.cpp
+	src/settings.cpp
```
Comment (JulianvDoorn) [inline Sources.cmake:32]: tabs vs spaces: spaces are preferred

File: TestModule.cmake [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+
+if (build_test_enabled)
+add_test (
+	NAME ${build_test}
+	COMMAND ${CMAKE_COMMAND} --build .
+)
+endif (build_test_enabled)
+
+if (unit_test_enabled)
+add_executable (${unit_test} ${unit_test_main})
```
Comment (JulianvDoorn) [inline TestModule.cmake:23]: Outdated project files.

It is advised to copy the template-arduino cmake files from https://github.com/R2D2-2018/Build-environment/tree/master/modules/template-arduino

File: src/coordinate3d.hpp [hunk 22]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @file
+ * @brief     This class contains the Coordinate3D Class
+ * @author    Jeroen van Hattem and Jeffrey de Waal
+ * @license   See LICENSE
```
Comment (JulianvDoorn) [inline src/coordinate3d.hpp:5]: Inconsistent license tag (see src/coordinate3d.cpp)

File: src/coordinate3d.cpp [hunk 23]
```diff
@@ -0,0 +1,34 @@
+/**
+ * @file
+ * @brief     Robot Arm class
+ * @author    Jeroen van Hattem and Jeffrey de Waal
+ * @license   MIT
```
Comment (JulianvDoorn) [inline src/coordinate3d.cpp:5]: Inconsistent license tag (see src/coordinate3d.hpp)

File: src/coordinate3d.hpp [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+ */
+#ifndef COORDINATE3D_HPP
+#define COORDINATE3D_HPP
+
+class Coordinate3D {
```
Comment (JulianvDoorn) [inline src/coordinate3d.hpp:10]: Class doxygen is preferred

File: src/queue.hpp [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+ * @license   See LICENSE
+ */
+
+#ifndef QUEUE_HPP
+#define QUEUE_HPP
+
+#include "wrap-hwlib.hpp"
+
+template <class T, size_t QUEUE_SIZE>
+class Queue {
```
Comment (JulianvDoorn) [inline src/queue.hpp:17]: Queue class doxygen missing @wvanbreukelen

File: src/robot_arm.hpp [hunk 26]
```diff
@@ -1,19 +1,132 @@
...
+#ifndef ROBOTARM_HPP
+#define ROBOTARM_HPP
 
+#include "coordinate3d.hpp"
+#include "uart_connection.hpp"
 #include "wrap-hwlib.hpp"
+#include <cstring>
 
-class robot_arm {
+class RobotArm {
```
Comment (JulianvDoorn) [inline src/robot_arm.hpp:15]: Class doxygen missing

File: src/settings.hpp [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+ * @brief   This class contains different settings to adjust the robot arm.
+ * @author  Jeroen van Hattem and Jeffrey de Waal
+ * @date    2018-05-17
+ */
+#ifndef SETTINGS_HPP
+#define SETTINGS_HPP
+
+#include "wrap-hwlib.hpp"
+
+class Settings {
```
Comment (JulianvDoorn) [inline src/settings.hpp:12]: Class doxygen missing

File: src/settings.hpp [hunk 28]
```diff
@@ -0,0 +1,39 @@
+/**
+ * @file    settings.hpp
+ * @brief   This class contains different settings to adjust the robot arm.
+ * @author  Jeroen van Hattem and Jeffrey de Waal
+ * @date    2018-05-17
+ */
```
Comment (JulianvDoorn) [inline src/settings.hpp:6]: License tag missing

File: src/robot_arm.hpp [hunk 29]
```diff
@@ -1,19 +1,132 @@
-#ifndef ROBOT_ARM_HPP
-#define ROBOT_ARM_HPP
+/**
+ * @file
+ * @brief     This class contains all the functions needed to use the robot arm
+ * @author    Jeroen van Hattem and Jeffrey de Waal
+ * @license   See LICENSE
```
Comment (JulianvDoorn) [inline src/robot_arm.hpp:5]: Inconsistent license tag

File: src/uart_connection.cpp (hunk 12 above)
Comment (JulianvDoorn) [reply src/uart_connection.cpp:87]: @wvanbreukelen

File: src/uart_connection.cpp (hunk 13 above)
Comment (JulianvDoorn) [reply src/uart_connection.cpp:90]: @wvanbreukelen

File: src/uart_connection.cpp (hunk 14 above)
Comment (JulianvDoorn) [reply src/uart_connection.cpp:112]: @wvanbreukelen

File: src/uart_connection.cpp (hunk 15 above)
Comment (JulianvDoorn) [reply src/uart_connection.cpp:119]: @wvanbreukelen

File: test/test_main.cpp [hunk 30]
```diff
@@ -0,0 +1,73 @@
+/**
+ * @file
+ * @brief     Test file
+ * @author    Jeroen van Hattem
+ * @license   MIT
+ */
+#define CATCH_CONFIG_MAIN // This tells Catch to provide a main() - only do this in one cpp file
+#include "../src/coordinate3d.hpp"
```
Comment (JulianvDoorn) [inline test/test_main.cpp:8]: using `../src` is disadvised. Most up-to-date project files allow including as `#include "coordinate3d.hpp"` directly

File: test/test_main.cpp [hunk 31]
```diff
@@ -0,0 +1,73 @@
...
+#include "../src/robot_arm.hpp"
```
Comment (JulianvDoorn) [inline test/test_main.cpp:9]: using `../src` is disadvised. Most up-to-date project files allow including as `#include "coordinate3d.hpp"` directly

File: test/test_main.cpp [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+
+#include "catch.hpp"
+
+RobotArm uarmSwiftPro;
+char text[100];
+
+TEST_CASE("Append char* with another char*") {
```
Comment (JulianvDoorn) [inline test/test_main.cpp:16]: For the small test cases, doxygen does not provide much information. But as test cases start to get really big, it is advised to write doxygen for them as well.

**Review by JulianvDoorn — CHANGES_REQUESTED** [review]

**Comment by JulianvDoorn** [comment]
Continuous Integration passed while it actually threw an error. This should be fixed before approval.

```
CMake Error at /usr/local/cmake-3.9.2/share/cmake-3.9/Modules/CMakeDetermineSystem.cmake:100 (message):
  Could not find toolchain file:
  /home/travis/build-environment/toolchain/compilers/host-gcc.cmake
Call Stack (most recent call first):
  CMakeLists.txt:13 (project)
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
```

**Pull request closed without merging**